### PR TITLE
Michael velo657

### PR DESCRIPTION
**Pull request by 0xMichaelRan**
Added redeem_stable_credit.go and basic input/abiInput unit test. 
Still need to finish the TODO funciton: TestRedeemStableCredit()

### CODE REVIEW

**Comment by 0xMichaelRan** [comment]
Resolved conflicts and run TestValidateRedeemStableCredit: 

```
GOROOT=/usr/local/go #gosetup
GOPATH=/Users/michael/go #gosetup
/usr/local/go/bin/go test -c -o /private/var/folders/h8/sj0jhj490xs749cttq3rnzqr0000gn/T/___TestValidateRedeemStableCredit_in_github_com_velo_protocol_DRSv2_go_libs_vclient github.com/velo-protocol/DRSv2/go/libs/vclient #gosetup
/usr/local/go/bin/go tool test2json -t /private/var/folders/h8/sj0jhj490xs749cttq3rnzqr0000gn/T/___TestValidateRedeemStableCredit_in_github_com_velo_protocol_DRSv2_go_libs_vclient -test.v -test.run ^TestValidateRedeemStableCredit$ #gosetup
=== RUN   TestValidateRedeemStableCredit
=== RUN   TestValidateRedeemStableCredit/success
=== RUN   TestValidateRedeemStableCredit/error,_validation_fail_invalid_collateralAmount_is_a_number_with_greater_than_7_decimal_places
=== RUN   TestValidateRedeemStableCredit/fail,_should_throw_error_invalid_RedeemAmount_format
=== RUN   TestValidateRedeemStableCredit/fail,_should_throw_error_RedeemAmount_must_be_positive
=== RUN   TestValidateRedeemStableCredit/fail,_should_throw_error_assetCode_must_not_be_blank
=== RUN   TestValidateRedeemStableCredit/fail,_should_throw_error_invalid_assetCode_format
--- PASS: TestValidateRedeemStableCredit (0.00s)
    --- PASS: TestValidateRedeemStableCredit/success (0.00s)
    --- PASS: TestValidateRedeemStableCredit/error,_validation_fail_invalid_collateralAmount_is_a_number_with_greater_than_7_decimal_places (0.00s)
    --- PASS: TestValidateRedeemStableCredit/fail,_should_throw_error_invalid_RedeemAmount_format (0.00s)
    --- PASS: TestValidateRedeemStableCredit/fail,_should_throw_error_RedeemAmount_must_be_positive (0.00s)
    --- PASS: TestValidateRedeemStableCredit/fail,_should_throw_error_assetCode_must_not_be_blank (0.00s)
    --- PASS: TestValidateRedeemStableCredit/fail,_should_throw_error_invalid_assetCode_format (0.00s)
PASS

Process finished with exit code 0
```

**Comment by 0xMichaelRan** [comment]
Add testing **TestClient_RedeemStableCredit**

```
GOROOT=/usr/local/go #gosetup
GOPATH=/Users/michael/go #gosetup
/usr/local/go/bin/go test -c -o /private/var/folders/h8/sj0jhj490xs749cttq3rnzqr0000gn/T/___redeem_stable_credit_test_go github.com/velo-protocol/DRSv2/go/libs/vclient #gosetup
/usr/local/go/bin/go tool test2json -t /private/var/folders/h8/sj0jhj490xs749cttq3rnzqr0000gn/T/___redeem_stable_credit_test_go -test.v -test.run "^TestValidateRedeemStableCredit|TestValidateRedeemStableCreditToAbiInput|TestClient_RedeemStableCredit$" #gosetup
=== RUN   TestValidateRedeemStableCredit
=== RUN   TestValidateRedeemStableCredit/success
=== RUN   TestValidateRedeemStableCredit/fail,_should_throw_error_invalid_RedeemAmount_format
=== RUN   TestValidateRedeemStableCredit/fail,_should_throw_error_RedeemAmount_must_be_positive
=== RUN   TestValidateRedeemStableCredit/fail,_should_throw_error_assetCode_must_not_be_blank
=== RUN   TestValidateRedeemStableCredit/fail,_should_throw_error_invalid_assetCode_format
--- PASS: TestValidateRedeemStableCredit (0.00s)
    --- PASS: TestValidateRedeemStableCredit/success (0.00s)
    --- PASS: TestValidateRedeemStableCredit/fail,_should_throw_error_invalid_RedeemAmount_format (0.00s)
    --- PASS: TestValidateRedeemStableCredit/fail,_should_throw_error_RedeemAmount_must_be_positive (0.00s)
    --- PASS: TestValidateRedeemStableCredit/fail,_should_throw_error_assetCode_must_not_be_blank (0.00s)
    --- PASS: TestValidateRedeemStableCredit/fail,_should_throw_error_invalid_assetCode_format (0.00s)
=== RUN   TestValidateRedeemStableCreditToAbiInput
=== RUN   TestValidateRedeemStableCreditToAbiInput/success
--- PASS: TestValidateRedeemStableCreditToAbiInput (0.00s)
    --- PASS: TestValidateRedeemStableCreditToAbiInput/success (0.00s)
=== RUN   TestClient_RedeemStableCredit
=== RUN   TestClient_RedeemStableCredit/Success
=== RUN   TestClient_RedeemStableCredit/error,_validation_fail
=== RUN   TestClient_RedeemStableCredit/error,_txHelper.ConfirmTx_returns_an_error
=== RUN   TestClient_RedeemStableCredit/error,_txHelper.ExtractRedeemEvent_returns_an_error
--- PASS: TestClient_RedeemStableCredit (0.01s)
    --- PASS: TestClient_RedeemStableCredit/Success (0.00s)
    --- PASS: TestClient_RedeemStableCredit/error,_validation_fail (0.00s)
    --- PASS: TestClient_RedeemStableCredit/error,_txHelper.ConfirmTx_returns_an_error (0.00s)
    --- PASS: TestClient_RedeemStableCredit/error,_txHelper.ExtractRedeemEvent_returns_an_error (0.00s)
PASS

Process finished with exit code 0

```

File: go/libs/vclient/redeem_stable_credit.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+	abiInput := input.ToAbiInput()
+	opt := bind.NewKeyedTransactor(&c.privateKey)
+	opt.GasLimit = constants.GasLimit
+
+	tx, err := c.contract.drs.Redeem(
+		opt,
+		abiInput.RedeemAmount,
+		abiInput.AssetCode,
+		)
```
Comment (jura-b) [inline go/libs/vclient/redeem_stable_credit.go]: don't forget to do go format

File: go/libs/vclient/redeem_stable_credit.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+type RedeemStableCreditInput struct {
+	RedeemAmount    string
+	AssetCode       string
+}
+
+func (i *RedeemStableCreditInput) Validate() error {
+	// validate RedeemAmount
+	if i.RedeemAmount == "" {
+		return errors.New("RedeemAmount must not be blank")
```
Comment (jura-b) [inline go/libs/vclient/redeem_stable_credit.go]: redeemAmount must not be blank (camel case)

File: go/libs/vclient/redeem_stable_credit.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	return &RedeemStableCreditAbiInput{
+		AssetCode:       i.AssetCode,
+		RedeemAmount:    redeemAmount,
+	}
+}
+
+type RedeemStableCreditOutput struct {
+	Tx      *types.Transaction
+	Receipt *types.Receipt
+	Event   *vabi.DigitalReserveSystemRedeem
```
Comment (jura-b) [inline go/libs/vclient/redeem_stable_credit.go]:
```
    ...
    Event *RedeemStableCreditEvent
    ...

type RedeemStableCreditEvent struct {
   // all fields must be string formatted
}
```

File: go/libs/vclient/tx_helper.go [hunk 4]
```diff
@@ -67,3 +67,18 @@ func (h *txHelper) ExtractMintEvent(eventName string, log *types.Log) (*vabi.Dig
...
+
+func (h *txHelper) ExtractRedeemEvent(eventName string, log *types.Log) (*vabi.DigitalReserveSystemRedeem, error) {
+	event := new(vabi.DigitalReserveSystemRedeem)
+	err := h.drsAbi.Unpack(event, eventName, log.Data)
+	if err != nil {
+		return nil, errors.Wrap(err, "fail to read event log")
+	}
+
+	// extract indexed field
+	if len(log.Topics) > 1 {
```
Comment (Kafakk) [inline go/libs/vclient/tx_helper.go]:
```suggestion
	if len(log.Topics) < 3 {
```

File: go/libs/vclient/tx_helper.go [hunk 5]
```diff
@@ -67,3 +67,18 @@ func (h *txHelper) ExtractMintEvent(eventName string, log *types.Log) (*vabi.Dig
...
+		return nil, errors.New("fail to parse indexed param of an event")
+	}
+	event.CollateralAssetCode = utils.BytesToBytes32(log.Topics[1].Bytes())
```
Comment (Kafakk) [inline go/libs/vclient/tx_helper.go]: add extract indexed field collateralAssetAddress

File: go/libs/vclient/redeem_stable_credit.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+	}
+
+	event, err := c.txHelper.ExtractRedeemEvent("Redeem", receipt.Logs[0])
+	if err != nil {
+		return nil, err
+	}
+	
+	redeemStableCreditEvent := &RedeemStableCreditEvent{
+		AssetCode:event.AssetCode,
+		StableCreditAmount:event.StableCreditAmount.String(),
```
Comment (Kafakk) [inline go/libs/vclient/redeem_stable_credit.go]:
```suggestion
		StableCreditAmount:utils.AmountToString(event.StableCreditAmount),
```

Comment (Kafakk) [reply go/libs/vclient/redeem_stable_credit.go]: convert to decimal format

File: go/libs/vclient/redeem_stable_credit.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+		CollateralAssetAddress:event.CollateralAssetAddress.String(),
+		CollateralAssetCode:string(event.CollateralAssetCode[:]),
```
Comment (Kafakk) [inline go/libs/vclient/redeem_stable_credit.go]:
```suggestion
		CollateralAssetCode:utils.Byte32ToString(event.CollateralAssetCode),
```

File: go/libs/vclient/redeem_stable_credit.go [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+	if !amount.IsPositive() {
+		return errors.New("redeemAmount must be positive")
+	}
+
+	// validate AssetCode
+	if len(i.AssetCode) == 0 {
+		return errors.New("assetCode must not be blank")
+	}
+
+	if matched, _ := regexp.MatchString(`^[A-Za-z0-9]{1,7}$`, i.AssetCode); !matched {
```
Comment (Kafakk) [inline go/libs/vclient/redeem_stable_credit.go]:
```suggestion
	if matched, _ := regexp.MatchString(`^[A-Za-z0-9]{1,12}$`, i.AssetCode); !matched {
```

Comment (Kafakk) [reply go/libs/vclient/redeem_stable_credit.go]: allow only 1-12 character

File: go/libs/vclient/redeem_stable_credit.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	redeemStableCreditEvent := &RedeemStableCreditEvent{
+		AssetCode:event.AssetCode,
+		StableCreditAmount:utils.AmountToString(event.StableCreditAmount),
+		CollateralAssetAddress:event.CollateralAssetAddress.String(),
+		CollateralAssetCode:utils.Byte32ToString(event.CollateralAssetCode),
+		CollateralAmount:event.CollateralAmount.String(),
```
Comment (Kafakk) [inline go/libs/vclient/redeem_stable_credit.go]:
```suggestion
		CollateralAmount:utils.AmountToString(event.CollateralAmount),
```

**Comment by 0xMichaelRan** [comment]
```
truffle(ganache)> let redeemStableCredit = await drs.redeem(16, "vUSD")
undefined
truffle(ganache)> redeemStableCredit
{ tx:
   '0xe1cfaf2c59747b149bbc3afb85902cc7073bbdd9766d9f8b6715b43bbadabb0e',
  receipt:
   { transactionHash:
      '0xe1cfaf2c59747b149bbc3afb85902cc7073bbdd9766d9f8b6715b43bbadabb0e',
     transactionIndex: 0,
     blockHash:
      '0x9819a169fed9b2f3771cb1a61008b7a43a1c839afb690dd2c8c92e965cec92d0',
     blockNumber: 66,
     from: '0xe256628887bfed3fc7e844273f1e4eafcc8d2365',
     to: '0xecd238eed0f156ce165230dd2a2c5a8bf79bd5a0',
     gasUsed: 171471,
     cumulativeGasUsed: 171471,
     contractAddress: null,
     logs: [ [Object], [Object] ],
     status: true,
     logsBloom:
      '0x0000000000000000000000000000000000000410000000000000000000000000000000000400000000000000200000000000000080010040000000000020000000040000000000000000000800000404000000000000000000000000000000000000000042000000004000000000080000000000000000000000001000000020000000000020010000000100000000000000000000000000000000000000000002000200004000400000000000000000000002200000800000000000800200000000000200000000000000000000000000c100000000000010000000000020000010000000000000000000000000040048000000000080000000000000000000',
     v: '0x1b',
     r:
      '0xc47a03e09e55e463a71caed39f68a3acb7f1139f3f0274a1ae15024797acce7',
     s:
      '0x1e91614886a85c6e77b097cf1e35fc96a3df14eb145ed0d81e11fddc46ef012d',
     rawLogs:
      [ [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object],
        [Object] ] },
  logs:
   [ { logIndex: 2,
       transactionIndex: 0,
       transactionHash:
        '0xe1cfaf2c59747b149bbc3afb85902cc7073bbdd9766d9f8b6715b43bbadabb0e',
       blockHash:
        '0x9819a169fed9b2f3771cb1a61008b7a43a1c839afb690dd2c8c92e965cec92d0',
       blockNumber: 66,
       address: '0xEcd238EEd0f156CE165230dd2a2C5A8BF79bd5a0',
       type: 'mined',
       id: 'log_6187c44c',
       event: 'Rebalance',
       args: [Result] },
     { logIndex: 6,
       transactionIndex: 0,
       transactionHash:
        '0xe1cfaf2c59747b149bbc3afb85902cc7073bbdd9766d9f8b6715b43bbadabb0e',
       blockHash:
        '0x9819a169fed9b2f3771cb1a61008b7a43a1c839afb690dd2c8c92e965cec92d0',
       blockNumber: 66,
       address: '0xEcd238EEd0f156CE165230dd2a2C5A8BF79bd5a0',
       type: 'mined',
       id: 'log_e4988c78',
       event: 'Redeem',
       args: [Result] } ] }
truffle(ganache)>
```

**Review by Kafakk — APPROVED** [review]

**Pull request merged**